DiffusionPreprocessing/scripts/run_eddy.sh

```diff
@@ -310,33 +310,49 @@ determine_eddy_tools_for_supported_six_series() {
-			if [ -e ${FSLDIR}/bin/eddy ]; then
-				# They have an ${FSLDIR}/bin/eddy. So use it.
-				g_gpuEnabledEddy="${FSLDIR}/bin/eddy"
-			elif [ -e ${FSLDIR}/bin/eddy_cuda ]; then
+			# However, it may still be possible to proceed, assuming
+			# that the user has created a symbolic link of either
+			# 'eddy_cuda' or (more dangerously) 'eddy' to the specific
+			# desired CUDA binary (e.g., ${FSLDIR}/bin/eddy_cuda -> eddy_cuda9.1)
+			if [ -e ${FSLDIR}/bin/eddy_cuda ]; then
 				# They have an ${FSLDIR}/bin/eddy_cuda. So use it.
 				g_gpuEnabledEddy="${FSLDIR}/bin/eddy_cuda"
+			elif [ -e ${FSLDIR}/bin/eddy ]; then
+				# They have an ${FSLDIR}/bin/eddy. It is dangerous to assume that this
+				# is a symlink to an eddy_cudaX.X version, esp. since recent FSL installers
+				# run an 'eddy_configuration.sh' script, that symlinks 'eddy' to 'eddy_openmp'
+				# (which is NOT a GPU/CUDA-enabled version). But we'll attempt to detect
+				# this situation below, and error out if it is detected.
+				g_gpuEnabledEddy="${FSLDIR}/bin/eddy"
 			else
-				# If they have neither an FSLDIR/bin/eddy or FSLDIR/bin/eddy_cuda,
+				# If neither an FSLDIR/bin/eddy_cuda or FSLDIR/bin/eddy exists,
 				# tell them that we can't figure out what eddy to use.
 				log_Err "Since you have requested the use of GPU-enabled eddy,"
-				log_Err "you must either have:"
-				log_Err "1. ${FSLDIR}/bin/eddy as a symbolic link to the version of"
-				log_Err "   eddy_cudaX.Y in ${FSLDIR}/bin that you want to use"
+				log_Err "you must either:"
+				log_Err "1. Set ${FSLDIR}/bin/eddy_cuda as a symbolic link to the version"
+				log_Err "   of eddy_cudaX.Y in ${FSLDIR}/bin that you want to use"
 				log_Err "   and is appropriate for the CUDA libraries installed"
 				log_Err "   on your system OR "
-				log_Err "2. ${FSLDIR}/bin/eddy_cuda as a symbolic link to the version"
-				log_Err "   of eddy_cudaX.Y in ${FSLDIR}/bin that you want to use OR"
-				log_Err "3. You must specify the --cuda-version=X.Y option to this "
+				log_Err "2. Specify the --cuda-version=X.Y option to this "
 				log_Err "   script in order to explicitly force the use of "
 				log_Err "   ${FSLDIR}/bin/eddy_cudaX.Y"
+				log_Err "3. Set ${FSLDIR}/bin/eddy as a symbolic link to the version of"
+				log_Err "   eddy_cudaX.Y in ${FSLDIR}/bin that you want to use"
+				log_Err "   (NOT RECOMMENDED since 'eddy' without any suffix is inherently ambiguous)"
 				log_Err_Abort ""
 			fi
+
+			# If g_stdEddy and g_gpuEnabledEddy are the same, we have a problem
+			# (probably a consequence of 'eddy' existing as a symlink to 'eddy_openmp')
+			if [ -e ${g_stdEddy} ]; then
+				# Check if files differ (which is what we want) within an 'if' statement
+				# so that we don't trigger any active error trapping if they do differ.
+				# 'diff' returns "true" if files are the same, in which case we want to abort.
+				# Don't wrap the 'diff' command in () or [], as that will likely change the behavior.
```

Comment:
FYI, `()` specifies a subshell (which does not behave like `$()`, which additionally captures stdout and returns it as a string), which in this case should do the same thing (same exit status as the last command it executed).  `(())` specifies a mathematical expression, with special syntax, `[]` uses a builtin `[` command that treats its arguments as a conditional expression (similar to the `test` builtin), and `[[]]` is bash syntax that behaves mostly like `[`, but with some enhancements (playing nicely with `&&` and `||` inside the expression).

`if` doesn't have any special syntax rules (it treats its arguments like any other bash command), those constructs can be used anywhere, it's just that `if` and `while` are usually the only things that need this kind of 'test' syntax.

Reply:
Should I not have merged this?

Reply:
The commits are fine, I was commenting on a comment, for explanatory purposes, not suggesting a change.

Reply:
I can remove the last line of the comment if you want (and just push the change directly to master), but are you saying that using `()`, `$()`, or `[]` would all survive the error trapping as well, in the case when the files differ?

Reply:
No.  Using `[]` would change the expected syntax (because `[` is particular about what its arguments are), and would throw an error.  Using `$()` would be wrong, as `if` would then try to execute whatever string came from stdout of the diff (but it is redirected currently, so it would always try to execute the empty string).

I am not suggesting a change, just trying to explain why `[]` would fail.  `if` itself doesn't have special syntax, but whatever you put *after* `if` generally does require special syntax.

Reply:
Maybe this is a less roundabout explanation: every `if [` statement was already using a nonzero exit status, whenever the condition was false, because a nonzero exit status was *always* the only way to tell `if` to skip the `then` section and do the `else` section if it exists - `[` was just the command that was giving the exit status:

```
tim@blackbox:~$ [ -e foo ]
tim@blackbox:~$ echo $?
1
```

Thus, nonzero exit status in an `if` or `while`, etc. is expected, and therefore these are specially exempted from the `trap ERR` and `set -e` rules.